examples/pipelines/commoncrawl/components/download_commoncrawl_segments/src/main.py

```diff
@@ -0,0 +1,77 @@
+import logging
+from typing import List
+
+import io
+import boto3
+
+import dask.bag as db
+import dask.dataframe as dd
+from dask.delayed import delayed
+
+import pandas as pd
+
+import requests
+from warcio.archiveiterator import ArchiveIterator
+
+from fondant.component import DaskTransformComponent
+
+logger = logging.getLogger(__name__)
+
+BASE_URL = "https://data.commoncrawl.org/"
+
+
+def get_records_from_warc_file(warc_file: str, n_records_to_download: int) -> List:
+    """Downloads a WARC file and extracts the webpages.
+    Args:
+        warc_file: The path to the WARC file.
+        n_records_to_download: The number of webpages to download from the WARC file.
+    Returns:
+        A list of webpages.
+    """
+    logger.info(f"Processing WARC file from segment path: {warc_file}...")
+    records = []
```

Comment:
do we risk running into out-of-memory issues if the number of extracted records is too large? right now it seems like we're collecting them all in a list and materializing them in memory before transforming them into a Dask Dataframe.

Reply:
Materializing a list and then converting it to a dataframe is the recommended approach, since appending a dataframe is a lot more expensive. We might want to do this on a partition level though.

Reply:
does that mean that we need one partition per WARC file to keep the dataframes as small as possible?

Reply:
I have tested this component and run into the expected out-of-memory. 
I have tried to download 1 segment with 30k records. Seems to be Philippes assumption is correct, that the `records` list didn't fit into memory in my case. 

What do you think about using `dask.delayed`? Basically, create a `delayed_object` with a fixed size (e.g. 1000 records), append them to a dask dataframe, and afterwards initialise the dataframe from the delayed objects. 

```
def load_record(record_counter, batch_size) -> list:
    # returning a list of delayed objects with a fixed size 
    offset = record_counter * batch_size # use this to determine the starting point to read from
    counter = 0
    records = []
    ...
    for record in ArchiveIterator(response.raw, arc2warc=True):
        if counter >= offset:
             # read content and append to list
        if len(records) >= batch_size:
             return records


delayed_data = []
for segment in segments:
    total_number_of_records = xxx
    batch_size = 1000
    delayed_data.extend([delayed(load_record)(record_counter, batch) for record_counter in range(0, 
    total_number_of_records, batch_size)])

# initialise the final dataframe, download will be executed
dataframe = dd.from_delayed(delayed_data)

```

Reply:
@mrchtr That's good idea, in fact Im also using chunking to create delayed objects for load_from_files component